src/sorting/counting_sort.rs

```diff
@@ -1,29 +1,64 @@
-pub fn counting_sort(array: &[usize]) -> Vec<usize> {
-    let max = array.iter().max().unwrap_or(&0);
+use std::collections::HashMap;
 
-    let mut count = vec![0; max + 1];
-    let mut output = vec![0; array.len()];
+pub fn counting_sort<T: Keyed + Ord + Copy>(array: &[T]) -> Vec<T> {
+    let mut new_hashmap: HashMap<usize, &T> = HashMap::new();
 
-    for &element in array.iter() {
+    for element in array {
+        new_hashmap.insert(element.key(), element);
+    }
+
+    let mut usize_array: Vec<usize> = Vec::new();
+
+    for element in array {
+        usize_array.push(element.key());
+    }
+
+    let max = usize_array.iter().max().unwrap_or(&0);
+    let mut count: Vec<usize> = vec![0; max + 1];
+    let mut output: Vec<T> = vec![array[0]; array.len()];
+
+    for &element in usize_array.iter() {
         count[element] += 1;
     }
 
     for i in 1..max + 1 {
         count[i] += count[i - 1];
     }
 
-    for i in (0..array.len()).rev() {
-        let j = array[i];
+    for i in (0..usize_array.len()).rev() {
+        let j = usize_array[i];
         count[j] -= 1;
-        output[count[j]] = array[i];
+        output[count[j]] = **new_hashmap.get(&usize_array[i]).unwrap();
```

Comment:
Whenever you see yourself having to write `**` explicitly there is something weird going on.

I think that the primary issue is having the `HashMap`. We aren't really taking advantage of the `key`.